File: src/Python/101-200/105.BuildTreeFromPreAndIn.py

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def buildTree(self, pre: List[int], infix: List[int]) -> TreeNode:
        n = len(pre)
        if n == 0:
            return []
        idx = {val:i for i,val in enumerate(infix)}
        st = []
        cur = root = TreeNode(pre[0])
        for i in range(1,n):
            st.append(cur)
            val = pre[i]
            if idx[val] < idx[cur.val]:
                cur.left = TreeNode(val)
                cur = cur.left
            else:
                while st and idx[val] > idx[st[-1].val]:
                    cur = st.pop()
                cur.right = TreeNode(val)
                cur = cur.right
        return root


if __name__ == "__main__":
    solution = Solution()
    print(solution.buildTree([3,9,20,15,7], [9,3,15,20,7]))
   

File: src/Java/101-200/105.BuildTreeFromPreAndIn.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

public class TreeNode {     
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = preorder.length;
        if (n == 0)
            return null;
        HashMap<Integer, Integer> idx = new HashMap<>();
        for (int i = 0; i < n; i++) {
            idx.put(inorder[i], i);
        }
        LinkedList<TreeNode> st = new LinkedList<>();
        TreeNode root = new TreeNode(preorder[0]);
        TreeNode cur = root;
        for (int i = 1; i < n; i++) {
            st.add(cur);
            int val = preorder[i];
            if (idx.get(val) < idx.get(cur.val)){
                cur.left = new TreeNode(val);
                cur = cur.left;
            }
            else{
                while (st.size() > 0 && idx.get(val) > idx.get(st.getLast().val)){
                    cur = st.removeLast();
                }           
                cur.right = new TreeNode(val);
                cur = cur.right;
            }
        }
        return root;
    }
}
